frontend/modules/shared/data-access/src/lib/services/ticket.service.ts

```diff
@@ -1,20 +1,41 @@
 import { Injectable } from '@angular/core';
 import { AbstractApiService } from './abstract-api.service';
 import { Ticket, TicketResponse } from '../models/ticket';
-import { Observable, of } from 'rxjs';
+import { Observable, Subject, combineLatest, map, switchMap } from 'rxjs';
+
+type TicketWithResponses = Ticket & { responses: TicketResponse[] };
 
 @Injectable({
   providedIn: 'root',
 })
-export class TicketService extends AbstractApiService<Ticket> {
+export class TicketService extends AbstractApiService<TicketWithResponses> {
   constructor() {
     super('/api/tickets');
   }
 
-  getResponses(id: Ticket['id']): Observable<TicketResponse[]> {
-    return this._http.get<TicketResponse[]>(
+  private readonly ticketSubject = new Subject<Ticket>();
+  private readonly responsesSubject = new Subject<TicketResponse[]>();
+
+  private fetchTicket(id: Ticket['id']) {
+    const refetch = super.get(id);
+    refetch.subscribe((ticket) => this.ticketSubject.next(ticket));
+    return refetch;
+  }
+
+  private fetchResponses(id: Ticket['id']) {
+    const refetch = this._http.get<TicketResponse[]>(
       `${this._resourceUrl}/${id}/responses`,
     );
+    refetch.subscribe((responses) => this.responsesSubject.next(responses));
```

Comment:
Jest jakiś konkretny cel robienia tutaj subscribe niż jak po prostu zwrócić http.get z tej funkcji?

Reply:
Obstawiam że brak wiedzy z Rxjs mnie to pokonał XD

Chciałem w tym serwisie funkcje do odświeżenia danych w observable który dostaje się z GET aby po jakimś PATCH albo POST dało się zakolejkować odświeżenie danych utrzymując nadal loading state.

Ten subscribe to sposób na wywołanie side effectu który zrobi refresh subjectów. Jak da się to zrobić bardziej po ludzku to daj znać XD

Reply:
To na pewno nie robiłbym tego w serwisie, bo teraz masz historie jednego ticketa w nim. Jakbyś wszedł bez przeładowania strony na innego ticketa to dostaniesz odpowiedzi z tego pierwszego.

Lepiej byłoby taką logikę odświeżania zawrzeć w komponencie, zamiast używania FormSubmitWrappera.
W ten sposób też mógłbyś się pozbyć dwóch formularzy, miałbyś stream np. coś takiego;

```ts
changeStatus$ = new Subject<Ticket['status']>();
changeStatusLoading$: Observable<bool> = this.changeStatus$.pipe(
  switchMap(status => this._service.changeStatus(status).pipe(
    // Tu można otworzyć powiadomienie, że sukces
    map(() => false), 
    startWith(true))
  ), 
  startWith(false),
  shareReplay(1)
);

_sendReponse$ = new Subject<string>();

sendResponse() {
 // Tu sprawdzenie poprawności formularza
  this._sendResponse$.next(this.form.controls.content.value);
}

ticket$ = this.changeStatusLoading$.pipe(
  filter(x => !x),
  switchMap(() => this._id$),
  switchMap(id => this._service.get(id).pipe(startWith(null)))
);

responses$ = this._sendResponse$.pipe(
  switchMap(content => this._service.send(content).pipe(
    // Tu można otworzyć powiadomienie, że sukces
    startWith(null))
  ),
  startWith(null),
  switchMap(() => this._id$),
  switchMap(id => this._service.getResponses(id))
);
```

```html
<tui-loader [showLoader]="changeStatusLoading$ | async">
  <!-- Tu przyciski wszystkie z np. (click)=changeStatus$.next('IRRELEVANT') -->
</tui-loader>
```

Na szybko napisane, możliwe, że są tu jakieś błędy, ale pewnie ogarniesz samą ideę. Jakbyś miał dalsze problemy z tym to możemy się zgadać

Reply:
Przeniosłem tą logikę do komponentu, muszę tylko jakoś jeszcze stan ładowania ogarnąć ale to I guess zapytam jutro na seminarium